### PR TITLE
Feature: Banning Users

### LINKED ISSUE #509: Dev: banning users
Implement a feature to ban users.

See
- https://forum.codeigniter.com/showthread.php?tid=83570&pid=401007#pid401007
- https://github.com/codeigniter4/shield/issues/459#issuecomment-1272450484

### PR DESCRIPTION
**Pull request by davidnsai**
Supersedes #617
Closes #509

The banning user feature has been completed. All recommendations from the previous pull request have been applied. @kenjis

- Signing Commits.
- Adding Tests.
- Creating a trait to check the status of the banned user unlike accessing the property directly.
- Adding Docs for the new feature.

### CODE REVIEW

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        $authConfig   = config('Auth');
+        $this->tables = $authConfig->tables;
+    }
+
+    public function up(): void
+    {
+        // Users Table
+        $fields = [
+            'banned'         => ['type' => 'tinyint', 'after' => 'active', 'constraint' => 1, 'null' => false, 'default' => 0],
+            'banned_message' => ['type' => 'varchar', 'after' => 'banned', 'constraint' => 255, 'null' => true],
```
Comment (datamweb) [inline src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: If `banned` and `banned_message` is going to be used instead of `status` and `status_message`, it is better, rename the previous field (`$forge->renameTable`) and use command (`$forge->modifyColumn()`) to change the properties.

This may seem like a **breaking change**, but I don't think it will bother many users. Because we have not been able to use `status` and `status_message` before.

This is my opinion, see what others say.

File: src/Filters/SessionAuth.php [hunk 2]
```diff
@@ -51,10 +51,18 @@ public function before(RequestInterface $request, $arguments = null)
 
             // Block inactive users when Email Activation is enabled
             $user = $authenticator->getUser();
+
+            if ($user->isBanned()) {
+                $authenticator->logout();
+
+                return redirect()->to(config('Auth')->logoutRedirect())
+                    ->with('error', lang('Auth.logOutBannedUser'));
```
Comment (datamweb) [inline src/Filters/SessionAuth.php:59]:
```suggestion
                $error = $user->getBanMessage() ?? lang('Auth.logOutBannedUser');
                $authenticator->logout();

                return redirect()->to(config('Auth')->logoutRedirect())
                    ->with('error', $error);
```
How about this?

File: src/Filters/SessionAuth.php [hunk 3]
```diff
@@ -51,10 +51,18 @@ public function before(RequestInterface $request, $arguments = null)
...
+            }
+
             if ($user !== null && ! $user->isActivated()) {
                 $authenticator->logout();
 
-                return redirect()->route('login')
+                return redirect()->to(config('Auth')->logoutRedirect())
```
Comment (datamweb) [inline src/Filters/SessionAuth.php:65]: Why did you change? Is this related to PR?

File: src/Language/fa/Auth.php [hunk 4]
```diff
@@ -7,6 +7,8 @@
     'unknownAuthenticator'  => '{0} احراز هویت معتبری نمی باشد.',
     'unknownUserProvider'   => 'قادر به تعیین ارائه دهنده کاربر برای استفاده نیست.',
     'invalidUser'           => 'قادر به پیداکردن کاربر مشخص شده نیست',
+    'bannedUser'            => '(To be translated) Can not log you in as you are currently banned.',
+    'logOutBannedUser'      => '(To be translated) You have been logged out because you have been banned.',
```
Comment (datamweb) [inline src/Language/fa/Auth.php:11]:
```suggestion
    'bannedUser'            => 'در حال حاضر نمی توانید وارد شوید، چون مسدود شده اید.',
    'logOutBannedUser'      => 'شما به دلیل مسدود شدن، از سیستم خارج شده اید.',
```

File: tests/Authorization/AuthorizableTest.php [hunk 5]
```diff
@@ -326,4 +326,24 @@ public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
         Locale::setDefault($currentLocale);
         Time::setTestNow();
     }
+
```
Comment (datamweb) [inline tests/Authorization/AuthorizableTest.php:329]: The test doesn't cover fully or at least 80%, however I don't mind.

**Review by datamweb — CHANGES_REQUESTED** [review]

File: docs/banning_users.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
+
+#  Banning Users
```
[the remark this answers is not in the document]
Comment (davidnsai) [reply docs/banning_users.md:2]: Noted, thanks, I'll be doing so as soon as I can

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php (hunk 1 above)
Comment (davidnsai) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: I was thinking the status can be utilised in the future for setting the status of a user. For example, a user may say set his status to away and then have a message to add context to the status that they set.

File: src/Filters/SessionAuth.php (hunk 3 above)
Comment (davidnsai) [reply src/Filters/SessionAuth.php:66]: No it isn't related. I changed it because I felt that a the route for login may not always be named that way and may be set in the configurations. Having the named of the route hardcoded may lead to 404 errors in future

Comment (datamweb) [reply src/Filters/SessionAuth.php:66]: OK, so it shouldn't be changed in this PR.
Please revert, if you think we need this change, please send in a new PR thread.

File: src/Language/ja/Auth.php [hunk 7]
```diff
@@ -7,6 +7,8 @@
     'unknownAuthenticator'  => '{0} は有効なオーセンティケーターではありません。', // '{0} is not a valid authenticator.',
     'unknownUserProvider'   => '使用するユーザープロバイダーを決定できません。', // 'Unable to determine the User Provider to use.',
     'invalidUser'           => '指定されたユーザーを見つけることができません。', // 'Unable to locate the specified user.',
+    'bannedUser'            => '(To be translated) Can not log you in as you are currently banned.',
+    'logOutBannedUser'      => '(To be translated) You have been logged out because you have been banned.',
```
Comment (kenjis) [inline src/Language/ja/Auth.php:11]:
```suggestion
    'bannedUser'            => '現在あなたはアクセスが禁止されているため、ログインできません。',
    'logOutBannedUser'      => 'アクセスが禁止されたため、ログアウトされました。',
```

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php (hunk 1 above)
Comment (kenjis) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: It seems Myth::Auth uses `status` for banning.
https://github.com/lonnieezell/myth-auth/blob/1c646c58e8b9b956b2163ebda8e5ec7e9ed609ce/src/Entities/User.php#L196-L202

Basically, Shield should use the minimum number of columns required.
We are hesitant to use the `status` because the use of the `status` is not clear.
If the `status` is the status of the user for administration, ban should use the `status`.

Comment (kenjis) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: @datamweb Renaming the column is a breaking change. If a dev already uses the `status` for some reason, the app will break.

But if a dev uses the `status` for user's current status (away, busy, etc.), then the app will break even if we use the `status` for banning.

After all, it is not good to have columns with unclear uses that are not being used.

File: src/Filters/SessionAuth.php (hunk 3 above)
Comment (davidnsai) [reply src/Filters/SessionAuth.php:66]: Noted, I'll revert then add that to another pull request.

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php (hunk 1 above)
Comment (jozefrebjak) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: @kenjis @davidnsai I would like to use `is_banned` rather than `banned` column with `tinyint` type. 

`banned_message` is for me extra column which is not need I rather to use existing `status` column for message. 

If user is banned then he just need to get message within login that he can't login because he was banned and his potentially set `status` like away etc. needs to be cleared after all.

Comment (kenjis) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: I thought about this.

Shield is an Authentication and Authorization library.

If the `status` is for user's current status (away, busy, etc.), it is not for Authentication,
nor for Authorization. We should not provide the `status`.

Essentially, banning is about authorization. So it can be implemented using the authorization function. 
However, authorization does not have the ability to provide (have) a message for a user in case of non-authorization.

It seems no problem that we use the `status` for the user status for users who can be authenticated but cannot login now in some reason. In this case, we can use the `status` for banning, account locking with too many login failures.

Comment (davidnsai) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: I think you are right there @kenjis. This means that the 2 new columns introduced for `banned` and `banned_message` are redundant but it now allows us to use the `status` field for a multitude of features without having to define a new column for every new feature that may be added in future. So in this case the `status` field would contain `banned` in a different scenario it may contain `locked` etc.

**Comment by lonnieezell** [comment]
@davidnsai @kenjis How close are we to getting this one in? Looks very close. I know we'd like to get a new release out, and I'm hoping to get this one in there. Can we make this happen?

File: src/Database/Migrations/2022-12-13-200557_add_ban_columns.php (hunk 1 above)
Comment (davidnsai) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: @lonnieezell I just need a resolution on whether I should get rid of the banned and banned_message fields and use the status and status_message fields then the feature should be ready.

Comment (kenjis) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: @lonnieezell See this thread.
I hope this PR uses the `status` column for banning like Myth::Auth.

Comment (lonnieezell) [reply src/Database/Migrations/2022-12-13-200557_add_ban_columns.php:32]: I had read this last night. I thought it was resolved to use the `status` column, which I agree with. That was the original intention for that column - to be used for a variety of system-related statuses, like banning, suspension, etc.

**Comment by lonnieezell** [comment]
@davidnsai Will you be able to make the change to use the `status` column? Would love to get this wrapped and get a new beta version out.

**Comment by davidnsai** [comment]
> @davidnsai Will you be able to make the change to use the `status` column? Would love to get this wrapped and get a new beta version out.

I'll be done with this today. I had a busy schedule lately.  The changes are not much

**Review by davidnsai — COMMENTED** [review]
These have all been resolved

File: src/Traits/Bannable.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        $users->save($this);
+
+        return $this;
+    }
+
+    /**
+     * Returns the ban message.
+     */
+    public function getBanMessage(): ?string
```
Comment (kenjis) [inline src/Traits/Bannable.php:54]: Can you add test for this method?

File: docs/banning_users.md [hunk 9]
```diff
@@ -0,0 +1,88 @@
+# Banning Users
+
+
```
Comment (kenjis) [inline docs/banning_users.md:3]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
+# Banning Users
+
+  
+
+Shield provides a way to ban users from your application. This is useful if you need to prevent a user from logging in, or logging them out in the event that they breach your terms of service.
+
+
```
Comment (kenjis) [inline docs/banning_users.md:7]: Same as above.

File: docs/banning_users.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
+# Banning Users
+
+  
+
+Shield provides a way to ban users from your application. This is useful if you need to prevent a user from logging in, or logging them out in the event that they breach your terms of service.
+
+  
+
+- [Checking if the User is Banned](#check-if-a-user-is-banned)
+
```
Comment (kenjis) [inline docs/banning_users.md:10]: Same as above.

File: docs/banning_users.md [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+- [Banning a User](#banning-a-user)
+
```
Comment (kenjis) [inline docs/banning_users.md:12]: Same as above.

File: docs/banning_users.md [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+- [Unbanning a User](#unbanning-a-user)
+
```
Comment (kenjis) [inline docs/banning_users.md:14]: Same as above.

File: docs/banning_users.md [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+- [Getting the Reason for Ban ](#getting-the-reason-for-ban)
+  
+
```
Comment (kenjis) [inline docs/banning_users.md:17]: Same as above.

File: docs/banning_users.md [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+- [Unbanning a User](#unbanning-a-user)
+
+- [Getting the Reason for Ban ](#getting-the-reason-for-ban)
+  
+ 
+
+###  Check if a User is Banned
+
+
```
Comment (kenjis) [inline docs/banning_users.md:21]: Same as above.

File: docs/banning_users.md [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+  
+ 
+
+###  Check if a User is Banned
+
+  
+
+You can check if a user is banned using `isBanned()` method on the `User` entity. The method returns a boolean `true`/`false`.
+
+
```
Comment (kenjis) [inline docs/banning_users.md:25]: Same as above.

File: docs/banning_users.md [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+You can check if a user is banned using `isBanned()` method on the `User` entity. The method returns a boolean `true`/`false`.
+
+  
+
+```php
+
+if ($user->isBanned()) {
+
+//...
```
Comment (kenjis) [inline docs/banning_users.md:31]:
```suggestion
    //...
```

File: docs/banning_users.md [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+###  Check if a User is Banned
+
+  
+
+You can check if a user is banned using `isBanned()` method on the `User` entity. The method returns a boolean `true`/`false`.
+
+  
+
+```php
+
```
Comment (kenjis) [inline docs/banning_users.md:28]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+if ($user->isBanned()) {
+
```
Comment (kenjis) [inline docs/banning_users.md:30]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+//...
+
```
Comment (kenjis) [inline docs/banning_users.md:32]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+}
+
```
Comment (kenjis) [inline docs/banning_users.md:34]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+
```
Comment (kenjis) [inline docs/banning_users.md:37]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+```
+
+  
+
+###  Banning a User
+
+
```
Comment (kenjis) [inline docs/banning_users.md:41]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+  
+
+###  Banning a User
+
+  
+
+To ban a user from the application, the `ban(?string $message = null)` method can be called on the `User` entity. The method takes an optional string as a parameter. The string acts as the reason for the ban.
+
+
```
Comment (kenjis) [inline docs/banning_users.md:45]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+  
+
+```php
+// banning a user without passing a message
+$user->ban();
+
+// banning a user with a message and reason for the ban passed.
+$user->ban('Your reason for banning the user here');
+
```
Comment (kenjis) [inline docs/banning_users.md:53]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+
```
Comment (kenjis) [inline docs/banning_users.md:56]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+  
+
+###  Unbanning a User
+
+  
+
+Unbanning a user can be done using the `unBan()` method on the `User` entity. This method will also reset the `status_message` property. 
+
+
```
Comment (kenjis) [inline docs/banning_users.md:64]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+###  Unbanning a User
+
+  
+
+Unbanning a user can be done using the `unBan()` method on the `User` entity. This method will also reset the `status_message` property. 
+
+  
+
+```php
+
```
Comment (kenjis) [inline docs/banning_users.md:67]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+$user->unBan();
+
```
Comment (kenjis) [inline docs/banning_users.md:69]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+
```
Comment (kenjis) [inline docs/banning_users.md:72]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+$user->unBan();
+
+```
+
+  
+
+###  Getting the Reason for Ban
+
+
```
Comment (kenjis) [inline docs/banning_users.md:76]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+  
+
+###  Getting the Reason for Ban
+
+  
+
+The reason for the ban can be obtained user the `getBanMessage()` method on the `User` entity.
+
+
```
Comment (kenjis) [inline docs/banning_users.md:80]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 33]
```diff
@@ -0,0 +1,88 @@
...
+###  Getting the Reason for Ban
+
+  
+
+The reason for the ban can be obtained user the `getBanMessage()` method on the `User` entity.
+
+  
+
+```php
+
```
Comment (kenjis) [inline docs/banning_users.md:83]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+$user->getBanMessage();
+
```
Comment (kenjis) [inline docs/banning_users.md:85]: Can you remove unnecessary empty lines?

File: docs/banning_users.md [hunk 35]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+
```
Comment (kenjis) [inline docs/banning_users.md:88]: Can you remove unnecessary empty lines?

File: src/Traits/Bannable.php (hunk 8 above)
Comment (davidnsai) [reply src/Traits/Bannable.php:54]: Done

File: docs/banning_users.md [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+You can check if a user is banned using `isBanned()` method on the `User` entity. The method returns a boolean `true`/`false`. 
+
+```php
+if ($user->isBanned()) {
+    //...
+}
+```  
+
+###  Banning a User
```
Comment (datamweb) [inline docs/banning_users.md:20]:
```suggestion
### Banning a User
```

File: docs/banning_users.md [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+To ban a user from the application, the `ban(?string $message = null)` method can be called on the `User` entity. The method takes an optional string as a parameter. The string acts as the reason for the ban.
+
+```php
+// banning a user without passing a message
+$user->ban();
+// banning a user with a message and reason for the ban passed.
+$user->ban('Your reason for banning the user here');
+```
+
+###  Unbanning a User
```
Comment (datamweb) [inline docs/banning_users.md:31]:
```suggestion
### Unbanning a User
```

**Review by datamweb — CHANGES_REQUESTED** [review]

**Review by datamweb — APPROVED** [review]
@davidnsai  Thank you! 🥀

@lonnieezell  I think this PR is ready to merge, if there is no case, please merge.

**Comment by lonnieezell** [comment]
@davidnsai Thanks for carrying this all of the way through! Merging.
